Add a guide for upgrading and migrating ES indices (#3321)
WordPress/openverse: documentation/ingestion_server/guides/migrate.md

# Index migration runbook

From time to time, we will need to update our Elasticsearch indices. These
modifications can be classified into two broad-strokes categories, depending on
whether the changes affect the main consumer of the indices, the API.

## Migration types

### API-free

These changes are safe modifications to the ES schema that do not affect the
API. As such they do not need any migration process. Examples:

- addition of new fields or subfields
- removal of fields that are not referenced or used by the API

> - changing the type to another compatible type (like `text` &harr; `keyword` )

Reviewer: This case is API-free only if the API does not refer to the field using the `.keyword` title (as we used to have for source, for example).

Author: No that's a misunderstanding because here `.keyword` was referring to the inner field which was confusingly named `keyword` (and had the type of `keyword`). 

Changing the type of a field from `text` to `keyword` does not entail any API changes and is API-free. It only changes how that field is indexed and searched internally by ES.

For API-free changes, we deploy the ingestion server and let a data-refresh to
occur. The indexes will be updated to the new schema without manual intervention
and will be made available to the API.

### API-involved

These changes are modifications to fields that already are in use by the API and
involve code changes in both the ingestion server and the API. Examples:

- removal of a field
- changing the type to an incompatible type
- renaming of a field

> Such kinds of changes need us to precisely deploy the API when the new index is

Reviewer: ```suggestion
Such kinds of changes need us to deploy the API right after the new index is
```
Clearer?

promoted because of these reasons:

- If we deploy a little late, the old field the API wants will disappear.
- If we deploy a little early, the new field the API wants will not be present.

This runbook documents guidelines and processes for API-involved migrations.

Our goal is to break down an API-involved change into multiple small, atomic
changes with each step affecting at most one of the ingestion server or the API
and ensuring that the API and ES remain compatible throughout the process.

## Pull request guidelines

A change that involves modification to the ES index as well as its usage in the
API requires at least three steps, each associated with exactly one PR that
modifies exactly one of the ingestion server or the API to allow them to be
deployed independently.

1. Change the ES index mapping in the ingestion server. Ensure that the change
   is purely additive, keeping the old fields unchanged and creating new fields
   that contain the data the API will need.

>    This PR should only make changes within the `ingestion_server/` directory.

Reviewer: ```suggestion
   This PR should make changes only within the `ingestion_server/` directory.
```

Author: Resolved in 128da0b.

> 2. Change the ES fields referenced by the API to use the new fields added in the

Reviewer: This might be a little unclear as "change the ES fields" sounds more like changing something in the Elasticsearch. Maybe something like "Update the API code to reference the new ES field added in the previous step"?

Author: Resolved in 128da0b.

   previous step. Ensure that the old fields become unreferenced.

   The PR should only make changes within the `api/` directory.

3. Change the ES index mapping in the ingestion server to remove the old,
   now-unreferenced fields.

>    This PR should only make changes within the `ingestion_server/` directory.

Reviewer: Maybe add the link (or paths) of the mapping files?

Author: Resolved in 128da0b.

```{tip}
Get the PRs reviewed in advance so that the entire process has been vetted by
the team and there are no surprises or delays when the plans have been set into
motion.
```

```{caution}
Each PR in the chain should branch from, and point to, its predecessor in the
chain so that CI continues to pass for each PR.
```

> ### Example

Reviewer: Really nice to have an example here!

Assume we have a field `foo` with type `text` in the index. It has a subfield
`keyword` with type `keyword`. The API uses `foo.keyword` for all purposes.

One PR that does these 3 things would be an API-involved change. So we split
them into 3 PRs.

> 1. Changing `foo` to type `keyword` would be an API-free change because it is a

Reviewer: _I think_ this is an API-free change because we are not using the `foo` field as a `keyword` field in the API, right? Or was it like only during our latest ES changes (we never used the `source` field in the API, only the `source.keyword`)

Author: Building on [this response](https://github.com/WordPress/openverse/pull/3321#discussion_r1385212742), a field with type `keyword` does not need the `.keyword` suffix, that suffix was a result of us using the a nested field named `keyword` inside the `source` field.

   type change between two compatible types and does not affect the nested field
   `foo.keyword` that is in use by the API. Technically the outer field can be
   assumed to be "new" because it was not being used at all.

2. Then we make an API change to use `foo` directly instead of `foo.keyword`.
   Any other accommodations to make use of `foo` can be made in this step. In
   this case `foo` will be the same as `foo.keyword` so no other changes will be
   needed.

3. Removal of the `foo.keyword` field would now also be an API-free change
   because the field would no longer be in use.

## Migration process

The entire migration process can be classified into 3 phases.

```{mermaid}
flowchart TD
  subgraph api[API]
    API
  end

  subgraph elasticsearch[Elasticsearch]
    image --> image-old
    image-filtered --> image-old-filtered
    audio --> audio-old
    audio-filtered --> audio-old-filtered
  end

  API --> image
  API --> image-filtered
  API --> audio
  API --> audio-filtered
```

> ### Create the new fields

Reviewer: I think we also need the rollback steps here.

Author: Step 2 inside this section points to the `upgrade.md` file which contains the forward steps as well as the rollback steps. Can you elaborate more to help me understand what change you are recommending?

Reviewer: Sorry I didn't realize that. It was much easier to understand this in the rendered docs :)

# Index migration runbook

From time to time, we will need to update our Elasticsearch indices. These
modifications can be classified into two broad-strokes categories, depending on
whether the changes affect the main consumer of the indices, the API.

## Migration types

### API-free

These changes are safe modifications to the ES schema that do not affect the
API. As such they do not need any migration process. Examples:

- addition of new fields or subfields
- removal of fields that are not referenced or used by the API
- changing the type to another compatible type (like `text` &harr; `keyword` )

> For API-free changes, we deploy the ingestion server and let a data-refresh to
> occur. The indexes will be updated to the new schema without manual intervention

Reviewer 2: ```suggestion
For API-free changes, we deploy the ingestion server and run a standard data-refresh
(either triggered manually or as scheduled). The indexes will be updated to the new schema without manual intervention
```

Author: Implemented in 66134fa.

and will be made available to the API.

### API-involved

These changes are modifications to fields that already are in use by the API and
involve code changes in both the ingestion server and the API. Examples:

- removal of a field
- changing the type to an incompatible type
- renaming of a field

> Such kinds of changes need us to precisely deploy the API simultaneously with

Reviewer 2: Since this doesn't _strictly_ have to happen simultaneously:
```suggestion
Such kinds of changes need us to precisely deploy the API in coordination with
```

Author: Implemented in 66134fa.

   - [`elasticsearch_models.py`](https://github.com/WordPress/openverse/tree/main/ingestion_server/ingestion_server/elasticsearch_models.py)

2. Update the API code to reference and use the new ES fields added in the
   previous step. Ensure that the old fields become unreferenced.

   The PR should make changes only within the `api/` directory.

3. Change the ES index mapping in the ingestion server to remove the old,
   now-unreferenced fields.

   Like PR number 1, this PR should also make changes only within the
   `ingestion_server/` directory.

```{tip}
Get the PRs reviewed in advance so that the entire process has been vetted by
the team and there are no surprises or delays when the plans have been set into
motion.
```

```{caution}
Each PR in the chain should branch from, and point to, its predecessor in the
chain so that CI continues to pass for each PR.
```

### Example

Assume we have a field `foo` with type `text` in the index. It has a subfield
`keyword` with type `keyword`. The API uses `foo.keyword` for all purposes.

One PR that does these 3 things would be an API-involved change. So we split

> them into 3 PRs.

Reviewer 2: What are the 3 things referenced in this line? Could you potentially reword this for clarity?

Author: Implemented in 66134fa.

1. Changing `foo` to type `keyword` would be an API-free change because it is a
   type change between two compatible types and does not affect the nested field
   `foo.keyword` that is in use by the API. Technically the outer field can be
   assumed to be "new" because it was not being used at all.

2. Then we make an API change to use `foo` directly instead of `foo.keyword`.
   Any other accommodations to make use of `foo` can be made in this step. In
   this case `foo` will be the same as `foo.keyword` so no other changes will be
   needed.

3. Removal of the `foo.keyword` field would now also be an API-free change
   because the field would no longer be in use.

## Migration process

The entire migration process can be classified into 3 phases.

> ```{mermaid}

Reviewer 2: These diagrams rock!!
